### ISSUE TITLE
Implement "Prepared Requests"

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2816]
## Problem

There are several steps we do for each request to become a generated query. e.g. for a GET:

https://github.com/PostgREST/postgrest/blob/1f13e43abec9fb7acda3967d3645d28e6a1f112a/src/PostgREST/App.hs#L167-L173

Before that:

https://github.com/PostgREST/postgrest/blob/1f13e43abec9fb7acda3967d3645d28e6a1f112a/src/PostgREST/App.hs#L148-L153

## Solution

Akin to "prepared statements", implement "prepared requests". We can make a hash of our input parameters and store the generated query. Then just execute the query using the arguments. 

## Notes

- Avoiding the ApiRequest step to parse the query parameters(`?id=eq.3&col=lt.4`) is not possible. If we switched to a grammar like `?id.eq=3&col.lt=4`(https://github.com/PostgREST/postgrest/issues/2066`) it might be possible to avoid our query grammar parsing and only do some lighter parsing.
- If we use the SEARCH method(https://github.com/PostgREST/postgrest/issues/2125#issuecomment-1376533758) and only have placeholder parameters on the query params then it should be possible to completely avoid parsing.   
- use a library like https://github.com/cocreature/lrucaching/
- check how postgresql does prepared statements to see if we can use similar modules
- similar idea to https://github.com/PostgREST/postgrest/issues/2698

**Comment by steve-chavez** [comment]
> hash only the "structure", so kind of like a "parametrized request", similar to the prepared statement?
> If not... then you'd need to parse way too much of the actual structure of the request before you could actually calculate the hash - you'd almost be there in that case.

Yes, only the structure. But as you mention that's not worth it. SEARCH looks good. 

My motivation is  that with https://github.com/PostgREST/postgrest/pull/2523 and https://github.com/PostgREST/postgrest/pull/1582 we're going to be putting more work at the Plan level. That will certainly not be zero cost in terms of perf.

---

https://github.com/PostgREST/postgrest/issues/2771 will prove how much time we take for parsing/planning.

**Comment by taimoorzaeem** [comment]
A cache for `QueryParams` sounds good. Maybe something like `LRU ByteString QueryParams` using [lrucache](https://hackage.haskell.org/package/lrucache-1.2.0.1/docs/Data-Cache-LRU.html)? I believe this would reduce parse time significantly for the same query.
